mathjax3-ts/output/svg/Notation.ts

```diff
@@ -0,0 +1,131 @@
+/*************************************************************
+ *
+ *  Copyright (c) 2018 The MathJax Consortium
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+/**
+ * @fileoverview  Implements utilities for notations for menclose elements
+ *
+ * @author dpvc@mathjax.org (Davide Cervone)
+ */
+
+import {BBox} from './BBox.js';
+import {SVGmenclose} from './Wrappers/menclose.js';
+import * as Notation from '../common/Notation.js';
+export * from '../common/Notation.js';
+
+/*******************************************************************/
+
+/**
+ * Shorthand for SVGmenclose
+ */
+export type Menclose = SVGmenclose<any, any, any>;
+
+/**
+ * The kinds of lines that can be drawn
+ */
+export type LineName = Notation.Side | ('vertical' | 'horizontal' | 'up' | 'down');
+
+/**
+ * The data for a given line as two endpoints: [x1, y1, x2, y1]
+ *
+ * @param {Menclose} node   The node whose line is to be drawn
+ * @param {LineName} line   The type of line to draw for the node
+ * @return {[number, number, number, number]}   The coordinates of the two nedpoints
+ */
+export const lineData = (node: Menclose, kind: LineName) => {
+    const {h, d, w} = node.getBBox();
+    const t = node.thickness / 2;
+    return ({
+        top: [0, h - t, w, h - t],
+        right: [w - t, -d, w - t, h],
+        bottom: [0, t - d, w, t - d],
+        left: [t, -d, t, h],
+        vertical: [w / 2 - t, h, w / 2 - t, -d],
+        horizontal: [0, (h - d) / 2 - t, w, (h - d) / 2 - t],
+        up: [t, t - d, w - t, h - t],
+        down: [t, h - t, w - t, t - d]
+    } as {[name: string]: [number, number, number, number]})[kind];
```

Comment:
Here's a lot of computation being done that is not necessary. In every call you need exactly to compute the one for `kind`. I'd suggest a table mapping `kind` to methods that of type `(number,  number, number, number): [number, number, number, number]`. And then have `return table[kind](h, d, w, t)`.

This way you can also avoid duplicate computations. In particular, divisions are costly.